File: apps/web/features/forms/utils/field-visibility.ts
import type {
  FieldVisibilityOperator,
  FieldVisibilityRule,
  FormField,
} from "@/features/forms/types";
import type { TaskFormResponses } from "@/features/tasks/types";

function normalizeComparable(value: string) {
  const normalized = value.trim().toLowerCase();
  if (normalized === "yes" || normalized === "ya" || normalized === "true" || normalized === "1") {
    return "yes";
  }
  if (
    normalized === "no" ||
    normalized === "tidak" ||
    normalized === "false" ||
    normalized === "0"
  ) {
    return "no";
  }
  if (normalized === "fail" || normalized === "gagal") {
    return "fail";
  }
  if (normalized === "pass" || normalized === "lulus") {
    return "pass";
  }
  return normalized;
}

export function resolveVisibilityRule(field: FormField): FieldVisibilityRule | null {
  const rule = field.options?.visibilityRule;
  if (rule?.fieldId) {
    return rule;
  }

  const legacyFieldId = field.options?.showWhenFieldId;
  if (!legacyFieldId) return null;

  return {
    fieldId: legacyFieldId,
    operator: "equals",
    value: field.options?.showWhenValue ?? "yes",
  };
}

function evaluateOperator(
  operator: FieldVisibilityOperator,
  actual: string,
  expected: string
): boolean {
  switch (operator) {
    case "is_empty":
      return actual.length === 0;
    case "is_not_empty":
      return actual.length > 0;
    case "equals":
      return normalizeComparable(actual) === normalizeComparable(expected);
    case "not_equals":
      return normalizeComparable(actual) !== normalizeComparable(expected);
    case "contains":
      return normalizeComparable(actual).includes(normalizeComparable(expected));
    default:
      return true;
  }
}

export function isFieldVisible(field: FormField, responses: TaskFormResponses): boolean {
  const rule = resolveVisibilityRule(field);
  if (!rule) return true;

  const actual = (responses[rule.fieldId] ?? "").trim();
  const expected = (rule.value ?? "").trim();
  const operator = rule.operator ?? "equals";

  if (operator === "is_empty" || operator === "is_not_empty") {
    return evaluateOperator(operator, actual, expected);
  }

  if (!actual && operator !== "equals" && operator !== "not_equals") {
    return false;
  }

  return evaluateOperator(operator, actual, expected);
}

export function getVisibleFields(fields: FormField[], responses: TaskFormResponses) {
  return fields.filter((field) => isFieldVisible(field, responses));
}

export const visibilityOperatorLabels: Record<FieldVisibilityOperator, string> = {
  equals: "sama dengan",
  not_equals: "tidak sama dengan",
  contains: "mengandung",
  is_empty: "kosong",
  is_not_empty: "tidak kosong",
};

File: apps/api/app/services/field_visibility.py
from __future__ import annotations

from typing import Any

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.models.form_field import FormField

YES_VALUES = {"yes", "ya", "true", "1"}
NO_VALUES = {"no", "tidak", "false", "0"}
FAIL_VALUES = {"fail", "gagal"}
PASS_VALUES = {"pass", "lulus"}


def _normalize_comparable(value: Any) -> str:
    normalized = str(value or "").strip().lower()
    if normalized in YES_VALUES:
        return "yes"
    if normalized in NO_VALUES:
        return "no"
    if normalized in FAIL_VALUES:
        return "fail"
    if normalized in PASS_VALUES:
        return "pass"
    return normalized


def _get_response(responses: dict[str, Any], field_id: str | int) -> str:
    key = str(field_id)
    if key in responses:
        return str(responses[key] or "").strip()
    if field_id in responses:
        return str(responses[field_id] or "").strip()
    return ""


def resolve_visibility_rule(options_json: Any) -> dict[str, Any] | None:
    if not isinstance(options_json, dict):
        return None

    rule = options_json.get("visibilityRule")
    if isinstance(rule, dict) and rule.get("fieldId"):
        return {
            "fieldId": str(rule["fieldId"]),
            "operator": str(rule.get("operator") or "equals"),
            "value": str(rule.get("value") or ""),
        }

    legacy_field_id = options_json.get("showWhenFieldId")
    if legacy_field_id:
        return {
            "fieldId": str(legacy_field_id),
            "operator": "equals",
            "value": str(options_json.get("showWhenValue") or "yes"),
        }

    return None


def is_field_visible(field: FormField, responses: dict[str, Any]) -> bool:
    rule = resolve_visibility_rule(field.options_json)
    if not rule:
        return True

    actual = _get_response(responses, rule["fieldId"])
    operator = rule.get("operator", "equals")
    expected = str(rule.get("value") or "")

    if operator == "is_empty":
        return len(actual) == 0
    if operator == "is_not_empty":
        return len(actual) > 0

    if not actual and operator not in {"equals", "not_equals"}:
        return False

    actual_norm = _normalize_comparable(actual)
    expected_norm = _normalize_comparable(expected)

    if operator == "equals":
        return actual_norm == expected_norm
    if operator == "not_equals":
        return actual_norm != expected_norm
    if operator == "contains":
        return expected_norm in actual_norm

    return True


def _is_filled(value: str) -> bool:
    return bool(value.strip())


def validate_conditional_required_fields(
    db: Session,
    *,
    form_template_id: int,
    responses: dict[str, Any],
) -> None:
    fields = (
        db.query(FormField)
        .filter(FormField.form_template_id == form_template_id)
        .order_by(FormField.sort_order.asc())
        .all()
    )

    missing: list[str] = []

    for field in fields:
        if not field.is_required:
            continue
        if not is_field_visible(field, responses):
            continue

        value = _get_response(responses, field.id)
        if not _is_filled(value):
            missing.append(field.label)

    if missing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "message": "Required conditional fields are missing.",
                "missing_fields": missing,
            },
        )
